include/sampleflow/consumers/average_cosinus.h

```diff
@@ -0,0 +1,205 @@
+// ---------------------------------------------------------------------
+
+#ifndef SAMPLEFLOW_CONSUMERS_AVERAGE_COSINUS_H
+#define SAMPLEFLOW_CONSUMERS_AVERAGE_COSINUS_H
+
+#include <sampleflow/consumer.h>
+#include <sampleflow/types.h>
+#include <mutex>
+
+#include <boost/numeric/ublas/matrix.hpp>
+
+
+namespace SampleFlow
+{
+  namespace Consumers
+  {
+  /**
+  * This is a Consumer class that implements computing the running average cosinus of
+  * angles between vectors value. This calculation, similarly to autocovariance, compares samples, that are
+  * apart by specific index. In this sense, we get function where domain is bounded unsigned int and for
+  * each value we calculate cosinus average.
+  * Let's say unsigned int is equal to l and n_samples=n and n is bigger than l. Than formula can be written as:
+ * $\hat{cos(\theta_n)(l)}=\frac{1}{n-l}\sum_{t=1}^{n-l}{(\bm{x}_{t+l}^T\bar\bm{x})(||bm{x}_{t}||*||\bar\bm{x}||)}.
+ *
+ * This code for every new sample updates $\hat{cos(\theta_n)(l)}, l=1,2,3...,cosinus_length. Choice of
+ * cosinus length can done by setting it in mcmc_test.cc.
+ *
+ * Notice, that for each new sample $x_n$, we need to take a sample, that was l earlier than new sample ($x_{n-l}).
+ * Working this way, updating algorithm becomes very similar to mean_value.h algorithm.
+ *
+ * Everytime, for updation we calculate corresponding fractions. There numerator is dot product, while denominator -
+ * multiplied those two vector norms
+ *
+ * ### Threading model ###
+ *
+ * The implementation of this class is thread-safe, i.e., its
+ * consume() member function can be called concurrently and from multiple
+ * threads.
+   **/
+
+    template <typename InputType>
+    class AverageCosinus: public Consumer<InputType>
+    {
+      public:
+        /**
+         * The data type of the elements of the input type.
+         */
+
+       using scalar_type = typename InputType::value_type;
+
+       /**
+   	 * The data type, where we save multiple set of past samples.
+   	 */
+
+       using value_type = boost::numeric::ublas::matrix<scalar_type>;
```

Comment:
The general convention is that the `value_type` is what is used as the return type of the `get()` function. If you need this type somewhere else, call it something different (and move the `using` declaration to the `private:` part of the class.